00_Base/src/services/credentials.service.ts

```diff
@@ -474,6 +477,119 @@ export class CredentialsService {
     return updatedClientInformation;
   }
 
+  async generateCredentialsTokenA(
+    credentialsRequest: CredentialsRequestDTO,
+    versionNumber: VersionNumber,
+  ): Promise<CredentialsDTO> {
+    const serverCredentialsRoleDTOs = credentialsRequest.roles;
+
+    for (const role of serverCredentialsRoleDTOs) {
+      try {
+        const storedRole =
+          await this.serverCredentialsRoleRepository.getServerCredentialsRoleByCountryCodeAndPartyId(
+            role.country_code,
+            role.party_id,
+          );
+        if (storedRole) {
+          throw new InternalServerError(
+            `ServerCredentialsRole with country_code ${role.country_code} and party_id ${role.party_id} already exists.`,
+          );
+        }
+      } catch (e) {
+        if (e instanceof NotFoundError) {
+          this.logger.debug(
+            `ServerCredentialsRole with country_code ${role.country_code} and party_id ${role.party_id} not found and can be created.`,
+          );
+        } else {
+          throw e;
+        }
+      }
+    }
+
+    // Make sure we stored the necessary version and version endpoints
+    // so that MSP can retrieve them later for registration process
+    const storedVersion = await this.versionRepository.readOnlyOneByQuery(
+      {
+        where: {
+          version: versionNumber,
+        },
+        include: [VersionEndpoint],
+      },
+      OcpiNamespace.Version,
+    );
+    if (!storedVersion) {
+      throw new NotFoundError(`Version not found for version ${versionNumber}`);
+    }
+    const credentialsUrl = storedVersion.endpoints.find(
+      (endpoint: any) => endpoint.identifier === ModuleId.Credentials,
+    );
+    if (!credentialsUrl) {
+      throw new NotFoundError(
+        `Endpoint URL for ${ModuleId.Credentials} not found for version ${versionNumber}`,
+      );
+    }
+
+    const cpoTenant = await CpoTenant.create(
+      {
+        serverCredentialsRoles: serverCredentialsRoleDTOs,
+      },
+      {
+        include: [
+          {
+            model: ServerCredentialsRole,
+            include: [
+              {
+                model: BusinessDetails,
+                include: [Image],
+              },
+            ],
+          },
+        ],
+      },
+    );
+    this.logger.info(`Created CpoTenant: ${JSON.stringify(cpoTenant)}`);
+
+    const credentialsTokenA = uuidv4();
+    // clientToken, clientCredentialsRoles and clientVersionDetails
+    // are added in the following post credentials requests based on the registration process
+    const clientInfo = await ClientInformation.create(
+      {
+        cpoTenantId: cpoTenant.id,
+        registered: false,
+        serverToken: credentialsTokenA,
+        serverVersionDetails: [
+          {
+            version: versionNumber,
+            url: credentialsRequest.url,
+            endpoints: storedVersion.endpoints.map((endpoint) => ({
```

Comment:
Hi @elliot-sabitov, in the discussion yesterday we decided not to add endpoints here. But then in the test I found after finishing the whole registration process, these endpoints were not get added. The later requests are focus on client versions. So I still added endpoints here. Please feel free let me know if I get it wrong. Thank you!